src/EditorFeatures/Core/ExternalAccess/EditorConfig/Features/Completion/CompletionHandler.cs

```diff
@@ -0,0 +1,191 @@
+            seenWhitespace = false;
+            foreach (var element in textToCheck)
+            {
+                if (seenWhitespace && !(element == ' '))
+                {
+                    return null;
+                }
+                seenWhitespace = element == ' ';
+            }
+
+            return CreateCompletionList(document, textInLine, showValueList: false);
+        }
+
+        private static CompletionItem? GenerateSettingNameCompletionItem(IEditorConfigSettingInfo setting)
+        {
+            var name = setting.GetSettingName();
+            if (name == null)
+            {
+                return null;
+            }
+            var documentation = setting.GetDocumentation();
+            var commitCharacters = _settingNameCommitCharacters;
+
+            return CreateCompletionItem(name, name, CompletionItemKind.Property, documentation, commitCharacters);
+        }
+
+        private static CompletionItem[]? GenerateSettingValuesCompletionItem(IEditorConfigSettingInfo setting, bool additional, OptionSet optionSet, bool allowsMultipleValues = false)
+        {
+            var values = new List<CompletionItem>();
+
+            // User may type a ',' but not in a setting that allows multiple values
+            if (additional && (!allowsMultipleValues))
+            {
+                return null;
+            }
+
+            // Create normal values list
+            var settingValues = setting.GetSettingValues(optionSet);
+            if (settingValues == null)
+            {
+                return null;
+            }
+
+            foreach (var value in settingValues)
+            {
+                var commitCharacters = allowsMultipleValues ? _multipleValuesCommitCharacters : ImmutableArray.Create(Array.Empty<string>());
+                values.Add(CreateCompletionItem(value, value, CompletionItemKind.Value, null, commitCharacters));
+            }
+
+            return values.ToArray();
+        }
+
+        private static CompletionItem CreateCompletionItem(string label, string insertText, CompletionItemKind kind, string? documentation, ImmutableArray<string> commitCharacters)
+        {
+            var item = new CompletionItem
+            {
+                Label = label,
+                InsertText = insertText,
+                Kind = kind,
+                Documentation = documentation,
+                CommitCharacters = commitCharacters.ToArray(),
+            };
+            return item;
+        }
+
+        private static CompletionItem[]? GetSettingValues(string settingName, ImmutableArray<IEditorConfigSettingInfo> settingsSnapshot, OptionSet optionSet, bool multipleValues = false)
+        {
+            var foundSetting = settingsSnapshot.Where(sett => sett.GetSettingName() == settingName);
+            if (foundSetting.Any())
+            {
+                var allowsMultipleValues = settingName == "csharp_new_line_before_open_brace" || settingName == "csharp_space_between_parentheses";
```

Comment:
is there any way to derive this from the setting or option itself?  Instead of harcoding.  If not, lets add a TODO to encode this information in the editorconfig value holder